bgpd/bgp_label.c

```diff
@@ -119,21 +119,137 @@ mpls_label_t bgp_adv_label(struct bgp_node *rn, struct bgp_path_info *pi,
 
 	return rn->local_label;
 }
+/**
+ * This is passed as the callback function to bgp_labelpool.c:bgp_lp_get()
+ * by bgp_reg_dereg_for_label() when a label needs to be obtained from
+ * label pool.
+ * Note that it will reject the allocated label if a label index is found,
+ * because the label index supposes predictable labels
+ */
+int bgp_reg_for_label_callback(
+	mpls_label_t new_label,
+	void *labelid,
+	bool allocated)
+{
+	struct bgp_path_info *pi = (struct bgp_path_info *)labelid;
+	struct bgp_node *rn = (struct bgp_node *)pi->net;
+	char addr[PREFIX_STRLEN];
+
+	prefix2str(&rn->p, addr, PREFIX_STRLEN);
+
+	zlog_debug("%s: FEC %s label=%u, allocated=%d",
+			__func__, addr, new_label, allocated);
+
+	if (!allocated) {
+		/*
+		 * previously-allocated label is now invalid
+		 */
+		if (pi->attr->label_index == MPLS_INVALID_LABEL_INDEX
+			&& pi->attr->label != MPLS_LABEL_NONE
+			&& CHECK_FLAG(rn->flags, BGP_NODE_REGISTERED_FOR_LABEL)
+		) {
+			bgp_unregister_for_label(rn);
+			label_ntop(MPLS_LABEL_IMPLICIT_NULL, 1,
+				   &rn->local_label);
+			bgp_set_valid_label(&rn->local_label);
+		}
+		return 0;
+	}
+
+	/*
+	 * label index is assigned, this should be handled by SR-related code,
+	 * so retry FEC registration and then reject label allocation for
+	 * it to be released to label pool
+	 */
+	if (pi->attr->label_index != MPLS_INVALID_LABEL_INDEX) {
+		zlog_debug("%s: FEC %s Rejecting allocated label %u as"
+				" Label Index is %u",
+				__func__, addr, new_label, pi->attr->label_index);
+
+		bgp_register_for_label(pi->net, pi);
+
+		return -1;
+	}
+
+	if (pi->attr->label != MPLS_INVALID_LABEL) {
+		if (new_label == pi->attr->label) {
+			/* already have same label, accept but do nothing */
+			return 0;
+		}
+		/* Shouldn't happen: different label allocation */
+		flog_err(EC_BGP_LABEL,
+			 "%s: %s had label %u but got new assignment %u",
+			 __func__, addr, pi->attr->label, new_label);
+		/* continue means use new one */
+	}
+
+	label_ntop(new_label, 1, &rn->local_label);
+	bgp_set_valid_label(&rn->local_label);
+
+	/*
+	 * Get back to registering the FEC
+	 */
+	bgp_register_for_label(pi->net, pi);
+
+	return 0;
+}
 
 void bgp_reg_dereg_for_label(struct bgp_node *rn, struct bgp_path_info *pi,
 			     int reg)
 {
+	int with_label_index = 0;
```

Comment:
The name here is "index." Today we only support one label (it seems, from the comments below), so it is only set to 0 or 1 today. But just to verify, this will be used for something other than 0 or 1 in the future, correct? Otherwise it might be better as a boolean (?).

Reply:
This variable is used as a boolean and is introduced for lighter code further on: to avoid having a lengthy condition twice (lines 223-224 and then the same check at line 265).

Reply:
Changed to bool.